src/replica/HttpAsyncClient.h

```diff
@@ -0,0 +1,312 @@
+ *       HttpAsyncClient::create(
+ *           [](auto const& reader) {
+ *               std::cout << "state: " << reader->state() << std::endl;
+ *           },
+ *           io_service,
+ *           "GET", "http://my.host.domain/data/chunk_0.txt");
+ *   });
+ *   reader->start();
+ *   io_service.run();
+ * @code
+ * 
+ * @note Once the start() method gets called w/o throwing an exception, the algorithm
+ *   can make repeated attempts to establish a connection to the Web server even if
+ *   the server's host and/or port won't be resolved, or if the connection attempts
+ *   will be failing, or if attempts to send a request will be failing. A short delay
+ *   will be introduced after each unsuccessfull attempt. The request will go into
+ *   the FAILED state if 1) there will be a failure in receiving server's response,
+ *   2) in case of incorrect data received from the server, or 3) the result is too
+ *   large.
+ *
+ * @note The implementation will open and close a new connection for each request.
+ * @note The implementation doesn't support TLS/SSL-based HTTPS protocol.
+ */
+class HttpAsyncClient: public std::enable_shared_from_this<HttpAsyncClient> {
+public:
+    /// The function type for notifications on the completion of the operation.
+    typedef std::function<void(std::shared_ptr<HttpAsyncClient> const&)> CallbackType;
+
+    enum class State: int {
+        CREATED = 0,        ///< The object was created and no request was sent
+        IN_PROGRESS,        ///< The request is still in progress
+        FINISHED,           ///< Final state: the operation succeeded, result is available 
+        FAILED,             ///< Final state: the operation failed
+        BODY_LIMIT_ERROR,   ///< Final state: the operation failed because the response's body is larger than requested
+        CANCELLED,          ///< Final state: the operation was explicitly cancelled after it was started
+        EXPIRED             ///< Final state: the operation was aborted due to the timeout expiration
+    };
+
+    /// @return The string representation of the state.
+    /// @throw std::invalid_argument For unknown values of the input parameter.
+    static std::string state2str(State state);
+
+    HttpAsyncClient() = delete;
+    HttpAsyncClient(HttpAsyncClient const&) = delete;
+    HttpAsyncClient& operator=(HttpAsyncClient const&) = delete;
+
+    /**
+     * @brief Static factory for creating objects of the class.
+     *
+     * Objects created by the method will be in the state State::CREATED. The actual
+     * execution of the request has to be initiated by calling the method start().
+     *
+     * @param io_service BOOST ASIO services.
+     * @param onFinish The callback function to be called upon a completion
+     *   (successfull or not) of a request.
+     * @param method The case-sensitive name of the HTTP method ('GET', 'POST', 'PUT', 'DELETE').
+     * @param url A location of the remote resoure.
```

Comment:
I need to parse URL by myself since we don't have a ready-to-use parser available in the current development container. There is an excellent Boost library that I would like to use https://master.url.cpp.al/. Though, it's not a part of the official Boost releases. The library is relatively new, and it requires some recent version of Boost that we don't have here. Ours is 1.66 (that was released circa 2017). The very same library could be used for parsing database connection strings. @fritzm is aware of the library.

Reply:
Upon furthe rinspection of the code, I have realized that I already have the class `replica::Url` that could be extended to suit needs of the async client. Class `replica::Url` alredy has a unit test. I will update the code to use `replica::Url` before we have a better solution, like the previously mentioned `Boost.Url`.